packages/okta-auth-js/lib/AsyncMethodQueue.js

```diff
@@ -0,0 +1,48 @@
+var util = require('./util');
+
+function AsyncMethodQueue() {
+  this.queue = [];
+  this.running = false;
+}
+
+util.extend(AsyncMethodQueue.prototype, {
+  push: function(/* method, thisObject, args... */) {
+    var args = Array.prototype.slice.call(arguments);
+    var method = args[0];
+    var thisObject = args[1];
+    args = args.slice(2);
+    return new Promise((resolve, reject) => {
+      this.queue.push({
+        method,
+        thisObject,
+        args,
+        resolve,
+        reject
+      });
+      this.run();
+    });
+  },
+  run: function() {
+    if (this.running) {
+      return;
+    }
+    if (this.queue.length === 0) {
+      return;
+    }
+    this.running = true;
+    var queueItem = this.queue.shift();
+    var res = queueItem.method.apply(queueItem.thisObject, queueItem.args);
+    if (util.isPromise(res)) {
```

Comment:
At this point, we are not directly exposing the queue to developers, so I don't think we will get any questions. Also I'm not sure we would want to support the callback format. We could provide another util to wrap a callback function into a Promise, but it's not a format we currently use in our SDK. I am mostly supporting sync functions here because it's cheap and easy. Not supporting and throwing an error would be a little more effort and seems unnecessary. The first time we encounter the error, it will probably be because we want to enqueue a sync function like `tokenManager.clear`, so we would remove the error logic and put it back to the way it is now.